ActionCable protocol negotiation

danielrhodes opened the pull request:
A first stab at negotiating an acceptable protocol as was mentioned in #23976 @maclover7 @jeremy @javan

Going forward, if the ActionCable protocol changes, there ought to be some way for the client/server to negotiate this change and gracefully stop communicating if no common protocol can be established.

Using the `Sec-Websocket-Protocol` from the [WebSocket spec](https://tools.ietf.org/html/rfc6455#page-59), the client and server negotiate a protocol from a list. Most of the work is being done by the [WebSocket driver](https://github.com/faye/websocket-driver-ruby/blob/master/lib/websocket/driver/hybi.rb#L76) and the web browser WebSocket client, so all this really does is make sure the client does not get into a bad state if the server has been upgraded and the client has not.

The browser websocket client automatically chooses to disconnect if the server does not respond with an acceptable protocol (this is not configurable), so instead of trying to reconnect (as this would not be useful), the client shuts down the monitor.

danielrhodes committed:
Added protocol negotiation

This is primarily for backwards compatibility for when
or if the protocol is changed in future versions.

If the server fails to respond with an acceptable
protocol, the client disconnects and disables
the monitor.

javan commented:
What browsers have you tested this in? I'm curious if protocols and `onclose` event codes are broadly supported. Unclear from documentation I've found online: https://developer.mozilla.org/en-US/docs/Web/API/WebSocket#Browser_compatibility and https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent#Browser_compatibility

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -46,6 +46,9 @@ class ActionCable.Connection
       @open()
 
+  protocol: ->
+    return @webSocket?.protocol
```
`return` is implicit in CoffeeScript, you can cut it.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -92,8 +95,15 @@ class ActionCable.Connection
+    close: (event) ->
       ActionCable.log("WebSocket onclose event")
+      if event.code == 1006
+        ActionCable.log("WebSocket received abnormal close error (" + event.code + "): " + event.reason)
```
More CoffeeScripty:
```coffee
if event.code is 1006
  ActionCable.log("WebSocket received abnormal close error (#{event.code}):", event.reason)
```

javan replied:
Also, why an additional log for `1006` only? Might as well log `event.code`, `event.reason`, and `event.wasClean` with every `close`.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -92,8 +95,15 @@ class ActionCable.Connection
+      if event.code == 1006
+        ActionCable.log("WebSocket received abnormal close error (" + event.code + "): " + event.reason)
+      
+      if !event.wasClean
```
In CoffeeScript you can do `unless event.wasClean` (or `if not event.wasClean`).

jeremy commented:
Cool! TIL that Sec-Websocket-Protocol is for subprotocol negotiation.

jeremy commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -92,8 +95,15 @@ class ActionCable.Connection
+      if !event.wasClean
+        ActionCable.log("WebSocket says close was unclean. Stopping ConnectionMonitor.")
+        @monitor.stop()
+
```
Can we bubble this up to the app in some way so it can react in the UI, e.g. show a prompt to reload the page or auto-reload it?

danielrhodes replied:
Yeah we need a way of telling the app why there was a hard disconnect. To be discussed, as there are other cases where we need this as well (e.g. authentication failure).

danielrhodes commented:
@javan Ok I tested it on Chrome, Safari, Firefox, and Mobile Safari. I don't have IE, but it does have the [`CloseEvent`](https://msdn.microsoft.com/en-us/library/hh772753(v=vs.85).aspx). Mobile Safari was the only one which behaved differently from the others (it did not automatically disconnect when the negotiation failed). Therefore, I changed the code to check the protocol negotiated to see if it is in the list the client knows it can accept and if not, to disconnect. I likewise took out that close event code stuff and checked the protocols as a way to decide if the monitor should be stopped. I think that should resolve your comments.

jeremy commented on actioncable/lib/action_cable.rb:
```diff
@@ -37,6 +37,10 @@ module ActionCable
+  
+  # This protocol ought to be versioned separately
+  # from the rest of the Gem
+  PROTOCOLS = ["actioncable-protocol-1.0.0".freeze].freeze
```
How about e.g. `actioncable-v17-json`? Tack on the transport encoding (which is always JSON now) and use a big chunky version that looks nothing like the AC version number since the only purpose is to break major compat (there are no "minor" changes).

Then the server can negotiate by parsing out the requested version and transfer encoding and picking the matching protocol handler for the conn.

matthewd replied:
Do we only want to version the ACa transport version, or also provide for an application-level value? I'm thinking of something equivalent to `config.assets.version`...

danielrhodes replied:
This is a good idea.

danielrhodes replied:
@matthewd This kind of gets to a bigger question of how much backwards-compatibility we build into this.

If somebody wants to make their ActionCable server a publicly available resource, then backwards compatibility becomes important. Likewise, I made a Swift client for ActionCable and a changing protocol version becomes a real pain for the developer because the longer upgrade cycle becomes a chicken-and-egg problem.

On the other hand, if this is really only used so a person can communicate with their own backend in the browser, there doesn't need to be a lot of backwards compatibility built-in. Only enough so that the client can gracefully degrade until the next reload.

In either case, there is no benefit for the developer in being able to pin a version because if the server is able to support a version, it should because it's an internal protocol.

I'm interested to see what feedback we get about this.

jeremy replied:
App-level protocol versioning is interesting. It could be helpful, at minimum, to provide a big hammer so apps can wiggle out of a stale-client compatibility mess by bumping the protocol.

For other cases, like changing channel behavior, I'd lean to leaving it up to apps to make changes in a compatible way. But, heck, leaning on bumping the app proto version could be simpler in many of these cases too.

jeremy commented on actioncable/lib/action_cable/connection/base.rb:
```diff
@@ -57,7 +57,9 @@ def initialize(server, env)
-        @websocket      = ActionCable::Connection::WebSocket.new(env, self, event_loop, server.config.client_socket_class)
+        @options = {:protocols => ActionCable::PROTOCOLS}
+
+        @websocket      = ActionCable::Connection::WebSocket.new(@env, @options, self, event_loop, server.config.client_socket_class)
```
Good one to pass as a keyword arg, no need to keep `@options` around

jeremy commented on actioncable/lib/action_cable/connection/client_socket.rb:
```diff
@@ -29,10 +29,11 @@ def self.secure_request?(env)
       attr_reader :env, :url
 
-      def initialize(env, event_target, event_loop)
+      def initialize(env, options, event_target, event_loop)
```
```ruby
def initialize(env, options, event_target, event_loop, protocols: ActionCable::PROTOCOLS)
```

jeremy replied:
Or glob the keyword args as websocket options:
```ruby
def initialize(env, options, event_target, event_loop, **websocket_options)
```
```ruby
@driver = ::WebSocket::Driver.rack(self, @websocket_options)
```

danielrhodes replied:
Would be happy to do this as it is cleaner. I haven't seen much use of the Ruby keyword arguments in the Rails codebase, but I guess it's ok now?

jeremy commented on actioncable/lib/action_cable/connection/client_socket.rb:
```diff
@@ -42,7 +43,7 @@ def initialize(env, event_target, event_loop)
 
         # The driver calls +env+, +url+, and +write+
-        @driver = ::WebSocket::Driver.rack(self)
+        @driver = ::WebSocket::Driver.rack(self, @options)
```
```ruby
@driver = ::WebSocket::Driver.rack(self, protocols: protocols)
```

danielrhodes commented:
In the comments @jeremy pointed out that it would be helpful if there was some way the client could communicate to the app a reason for the disconnect. There isn't currently a global way of doing this, so any opinions on this? We can pass some sort of `reason` object through to the disconnect method like `WebSocket` does.

jeremy commented on actioncable/lib/action_cable/connection/faye_client_socket.rb:
```diff
@@ -3,9 +3,10 @@
+      def initialize(env, options, event_target, stream_event_loop)
         @env = env
         @event_target = event_target
+        @options = options || {}
```
Ditto

jeremy commented on actioncable/lib/action_cable/connection/faye_client_socket.rb:
```diff
@@ -31,7 +36,9 @@ def rack_response
-          @faye = Faye::WebSocket.new(@env)
+
+          protocols = @options[:protocols] || nil
+          @faye = Faye::WebSocket.new(@env, protocols)
```
Do we need the `|| nil`? Would only kick in if protocols was oddly `false`

jeremy commented on actioncable/lib/action_cable/connection/web_socket.rb:
```diff
@@ -4,8 +4,8 @@ module ActionCable
-        @websocket = ::WebSocket::Driver.websocket?(env) ? client_socket_class.new(env, event_target, event_loop) : nil
+      def initialize(env, options, event_target, event_loop, client_socket_class)
+        @websocket = ::WebSocket::Driver.websocket?(env) ? client_socket_class.new(env, options, event_target, event_loop) : nil
       end
```
Ditto

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,11 +92,20 @@ class ActionCable.Connection
       ActionCable.log("WebSocket onclose event")
+      if @webSocket?.protocol not in ActionCable.PROTOCOLS
+        ActionCable.log("WebSocket is closing because there is no common protocol. Stopping ConnectionMonitor.")
+        @monitor.stop()        
```
This will prevent the connection monitor from infinitely attempting to reconnect (as noted in https://github.com/rails/rails/pull/23976#issuecomment-197627120), but leaves you disconnected until you reload your browser. Think we should notify subscriptions of this state so they can manage / instruct reloading.

danielrhodes replied:
Yes, @jeremy wants something along these lines. I guess just a simple object with `code` and a reason?

javan replied:
Yeah, I think that'd work fine. Would be cool to hot-reload the client-side js, but that's probably more trouble than it's worth for what will be an infrequent event.

danielrhodes replied:
I guess if you could plug into Sprockets somehow, you could pull a new version of the classes (perhaps namespaced differently for each asset build) and then hot swap in the ActionCable classes and reconnect. It would be super cool, but yes a lot of trouble.

As a side note: Facebook was doing [something similar](http://developer.android.com/reference/dalvik/system/DexClassLoader.html) on Android for awhile, before Google made them stop by modifying the TOS for the Google Play Store.

javan commented on actioncable/app/assets/javascripts/action_cable.coffee.erb:
```diff
@@ -3,6 +3,7 @@
 
 @ActionCable =
   INTERNAL: <%= ActionCable::INTERNAL.to_json %>
+  PROTOCOLS: <%= ActionCable::PROTOCOLS.to_json %>
```
How about adding `:protocols` to `INTERNAL` instead? It's a general purpose data bag.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,11 +92,20 @@ class ActionCable.Connection
+      # This is in the case that the browser client wants
+      # to open the connection regardless of whether a
+      # protocol was negotiated successfully.
+      if @webSocket?.protocol not in ActionCable.PROTOCOLS
```
I'd extract this to a new method since it's duplicated

maclover7 commented on actioncable/lib/action_cable.rb:
```diff
@@ -37,6 +37,10 @@ module ActionCable
+  
+  # This protocol ought to be versioned separately
+  # from the rest of the Gem
+  PROTOCOLS = ["actioncable-v1-json".freeze].freeze
```
Do we want this to have the same version number as the gem? Would prevent having two different version numbers for Action Cable...

jeremy replied:
The protocol version is distinct from the library version and will be stable over longer periods. Using separate versions is intentional.

jeremy commented on actioncable/lib/action_cable/connection/faye_client_socket.rb:
```diff
@@ -31,7 +36,9 @@ def rack_response
           return if @faye
-          @faye = Faye::WebSocket.new(@env)
+
+          protocols = @options[:protocols] || nil
```
Missed this line? options gone now

danielrhodes replied:
Oops  :-)

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,11 +92,20 @@ class ActionCable.Connection
+    close: (event) ->
       ActionCable.log("WebSocket onclose event")
+      if @webSocket?.protocol not in ActionCable.PROTOCOLS
+        ActionCable.log("WebSocket is closing because there is no common protocol. Stopping ConnectionMonitor.")
```
For debugging, it would be useful to log the client and server protocol versions rather than say they're not common.

danielrhodes replied:
The server does not tell the client which subprotocols it supports. It simply chooses one or none, and then responds with that protocol in the header.

javan replied:
I mean add `@webSocket.protocol` and `ActionCable.PROTOCOLS` to the log message

danielrhodes commented:
@javan I removed `@disconnect()` from the `error` handler for the websocket. From what I understand, this always gets fired with `closed` anyways, but it creates a race condition and doesn't provide any information.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -108,15 +110,17 @@ class ActionCable.Connection
-      @disconnect()
+        @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
+      else
+        @disconnect(new ActionCable.Events.Disconnected(event.code, event.reason, true))
```
Subscriptions don't receive Event objects anywhere else so I think passing a plain object would do. Perhaps:
```js
 `{ code: 123, message: "...", willReconnect: (true|false) }
```
I think `willReconnect` better indicates wether action is needed.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -1,9 +1,11 @@
 
 {message_types, protocols} = ActionCable.INTERNAL
 
+
```
:scissors:

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -108,15 +110,17 @@ class ActionCable.Connection
-    return if @disconnected
+  disconnect: (event) ->
+    if event && !event.isRecoverable
+      @monitor.stop()
```
Think this should move to the `close` handler so it's more closely tied to the related condition.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -46,6 +48,12 @@ class ActionCable.Connection
+    @webSocket?.protocol
+
+  isProtocolSupported: ->
+    @webSocket?.protocol in protocols
```
Can use the `protocol` method instead of repeating `@webSocket?.protocol`

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -23,9 +25,9 @@ class ActionCable.Connection
       throw new Error("Existing connection must be closed before opening")
     else
-      ActionCable.log("Opening WebSocket, current state is #{@getState()}")
+      ActionCable.log("Opening WebSocket, current state is #{@getState()}, subprotocols: " + protocols.join(","))
```
Arrays are implicitly joined with `,` when coerced to a String. Can do `.log("Opening WebSocket, current state is #{@getState()}, subprotocols: #{protocols}")`

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,19 +97,31 @@ class ActionCable.Connection
+      if not @isProtocolSupported()
+        @close()
+      else
+        ActionCable.log("WebSocket using '" + @protocol() + "' subprotocol.")
```
I'd move this logging to the initial `open` entry. And prefer CoffeeScript's string interpolation: `"using #{@protocol()}  subprotocol"`

danielrhodes replied:
There is a log in the initial open which shows the available protocols. The log above is showing the actual protocol chosen by the server.

javan replied:
I'm suggesting combining them. No need for two log entries.
```coffee
 open: ->
   ActionCable.log("WebSocket onopen event, subprotocol is #{@protocol()}")
```

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -46,6 +48,12 @@ class ActionCable.Connection
       @open()
 
+  protocol: ->
+    @webSocket?.protocol
```
`getProtocol` would better match the other accessor methods like `getState`.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -46,6 +48,12 @@ class ActionCable.Connection
+  protocol: ->
+    @webSocket?.protocol
+
+  isProtocolSupported: ->
```
ditto re: matching method naming convention: `isSupportedProtocol`

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,19 +97,31 @@ class ActionCable.Connection
+      # This is in the case that the browser client opens
+      # the connection regardless of whether a protocol was 
+      # negotiated successfully or not.
+      if not @isProtocolSupported()
```
Might as well flip the conditions to avoid `if not`ing.
```coffee
if @isProtocolSupported()
  ...
else
  @close()
```

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,19 +97,31 @@ class ActionCable.Connection
+    close: (event) ->
       ActionCable.log("WebSocket onclose event")
-      @disconnect()
+      if not @isProtocolSupported()
```
ditto re: flipping condition

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,19 +97,31 @@ class ActionCable.Connection
 
     error: ->
       ActionCable.log("WebSocket onerror event")
-      @disconnect()
```
>  I removed `@disconnect()` from the `error` handler for the websocket. From what I understand, this always gets fired with `close` anyways, but it creates a race condition and doesn't provide any information.

This assumes every `error` results in a `close`, yeah? If you can verify then :+1:, but otherwise it makes me a little uncomfortable.

danielrhodes replied:
https://www.w3.org/TR/websockets/#concept-websocket-close-fail

> If the user agent was required to fail the WebSocket connection or the WebSocket connection is closed with prejudice, fire a simple event named error at the WebSocket object.

Firefox and Safari do not consistently fire this `error` event at all, but they all seem to fire the `close` event consistently.

maclover7 commented on actioncable/lib/action_cable.rb:
```diff
@@ -35,7 +35,8 @@ module ActionCable
     },
-    default_mount_path: '/cable'.freeze
+    default_mount_path: '/cable'.freeze,
+    protocols: ["actioncable-v1-json".freeze].freeze
```
Isn't there only supposed to be one protocol here? Not sure why we need an array :grimacing:

danielrhodes replied:
The websocket on both ends takes an array and for potential backwards compatibility reasons, it may be necessary in the future for the server to be able to respond to multiple versions (cross your fingers this doesn't come true!).

maclover7 commented on actioncable/lib/action_cable/connection/faye_client_socket.rb:
```diff
@@ -3,9 +3,10 @@
   module Connection
     class FayeClientSocket
-      def initialize(env, event_target, stream_event_loop)
+      def initialize(env, event_target, stream_event_loop, protocols: ActionCable::INTERNAL[:protocols])
```
I think we can drop the default here for protocols -- the option to pass in a different array is already presented to users in `Connection::WebSocket#initialize`

maclover7 commented on actioncable/lib/action_cable/connection/client_socket.rb:
```diff
@@ -29,7 +29,7 @@ def self.secure_request?(env)
       attr_reader :env, :url
 
-      def initialize(env, event_target, event_loop)
+      def initialize(env, event_target, event_loop, protocols: ActionCable::INTERNAL[:protocols])
```
ditto -- I think we can drop the default here for protocols -- the option to pass in a different array is already presented to users in Connection::WebSocket#initialize

danielrhodes commented:
Just had a bunch of fun trying to figure out an issue where, when the browser attempts to reconnect on a failed connection, it can sometimes look the same as if there was a protocol mismatch. To my horror this was also specified as the [correct implementation of websockets](https://www.w3.org/TR/websockets/#concept-websocket-close-fail). That would leave us with no ability to stop retrying/tell the app that the connection was not going to work.

Luckily this only seems to happen in Chrome. My solution, which is a tad bit hacky, is to test the connection using a temporary websocket connection but without protocols, and if that goes through, we know the protocols we tried before were incorrect. Tested on Chrome, Safari, Mobile Safari, and Firefox.

Does somebody have the ability to test on IE?

maclover7 commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,19 +97,55 @@ class ActionCable.Connection
-    @subscriptions.notifyAll("disconnected")
+    @subscriptions.notifyAll("disconnected", event)
     @monitor.recordDisconnect()
+
```
Bunch of empty lines here...?

maclover7 commented on actioncable/app/assets/javascripts/action_cable/events.coffee:
```diff
@@ -0,0 +1,12 @@
+class ActionCable.Events extends Event
+    #
```
Do we need these commented out lines?

maclover7 commented on actioncable/app/assets/javascripts/action_cable/events.coffee:
```diff
@@ -0,0 +1,12 @@
+class ActionCable.Events.Disconnected extends ActionCable.Events
+  
+  constructor: (@code, @reason, @isRecoverable) ->
+    #
```
ditto - Do we need these commented out lines?

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,19 +97,50 @@ class ActionCable.Connection
+
+      # This encompases the behavior of most browsers when it comes to
+      # reacting to a failed protocol negotiation.
+      if event.code == 1000 && event.wasClean && @getProtocol() == ""
```
Prefer CoffeeScript's operators: `if event.code is 1000 and event.wasClean and @getProtocol() is ""`

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,19 +97,50 @@ class ActionCable.Connection
+      # exactly the same as a normal bad disconnect. Thus, we must test to 
+      # see if the server would connect under normal conditions. If so, Chrome reacted
+      # to the failed protocol negotiation, and not something else.
+      else if event.code == 1006 && !event.wasClean && @getProtocol() == ""
```
ditto re: operators

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,19 +97,50 @@ class ActionCable.Connection
+        connection = this
+        testSocket = new WebSocket(@consumer.url)
+        testSocket.onopen = ->
+          connection.disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
```
You can use a fat arrow function, which binds to `this`:
```coffee
testSocket.onopen = =>
  @disconnect(...)
```

danielrhodes replied:
Will give this a try. Thanks for the tips! My CoffeeScript skillz are nascent.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,19 +97,49 @@ class ActionCable.Connection
+        testSocket = new WebSocket(@consumer.url)
+        testSocket.onopen = =>
+          @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
+          testSocket.close()
```
Hmm, I don't think opening a new connection here is a good idea. Imagine a scenario when you're performing a server restart and suddenly every user attempts to reconnect. A flood of traffic like that can be crippling. I'm not totally clear on the circumstances that lead to this condition though so maybe you can elaborate.

danielrhodes replied:
So this is all so we can give the app a message saying it was disconnected for good and why.

I'm not a big fan of doing this either, but there isn't a good alternative for Chrome. The reason for all this is that Chrome does not distinguish between a bad disconnect and a protocol mismatch. In either case, it disconnects. For some reason the standard for websockets in browsers wants to disincentivize "probing" a websocket for capabilities (as if you would do this from a web browser!).

Thus, the only way to reliably tell if Chrome disconnected because of your normal bad connection is to create a new connection without protocols to see if Chrome stays connected.

This should only occur in two circumstances:
1. A previous connection attempt on Chrome was unsuccessful at connecting, at which point the server is down anyways.
2. There was a protocol mismatch when the server was up, at which point Chrome is going to disconnect, reconnect again without protocols, see that was the reason, and then disconnect permanently.

The alternative to this is just not upgrading the connection when the server sees that the client wants protocols it doesn't have. I don't think we could get an error code off that.

danielrhodes replied:
By the way, that thundering herd issue is still going to be a problem with the current retry handler. Given that all clients disconnect at the same time from a restarting server, they will all be on similar intervals when retrying. Could solve this by adding some jitter.

javan replied:
That would only be true if every client connected at the exact same time, which will never be the case. `ConnectionMonitor`'s timer naturally distribute the herd based on the randomness of when each client opened a connection.

javan replied:
> Thus, the only way to reliably tell if Chrome disconnected because of your normal bad connection is to create a new connection without protocols to see if Chrome stays connected.

If we can't reliably read the close event details, perhaps we should instead send a special rejection message that instructs the client to disconnect.

danielrhodes replied:
Well we can't send a `Sec-Websocket-Protocol` and then not receive one back from the server: the browser websocket clients just force disconnect. So that would mean moving the negotiation to outside of the websocket layer, which involves changing the protocol, and itself becomes subject to changes. :-)

Closing the connection is no problem: the browser is forcefully doing that for us. This whole thing is quite effective in preventing the client side from getting into a bad state, which is the primary concern.

But the issue that remains is we can't tell the app why there has been this specific disconnect in Chrome, unless we use this reconnect thing. For me, that's not ideal but it's not a deal breaker.

I see a few options:

1.  Keep the double reconnection thing. It's going to result in Chrome doing two very short-lived connections, but only if the protocol changes. However, the connection attempts will stop after that.
   **Pros**: We get consistent behavior across all browsers; from a load perspective it will in aggregate be less (no retries).

  **Cons**: _Only_ when there is a protocol change, there will be two back to back connection attempts by Chrome users; Unclear what load profile this would create. Solution is non standard.
2. Let Chrome continue to retry as it normally would.
  **Pros**: No back to back connection attempt for Chrome.

  **Cons**: Inconsistent behavior across browsers, and Chrome users won't be able to get a user-friendly message to refresh. Chrome will continue to retry fruitlessly.
3. Don't attempt to understand reason for connection and continue retrying, but the app won't know why it disconnected.
  **Pros**: Consistent across all browsers.

  **Cons**: No nice message to users. All browsers are going to continue to retry fruitlessly.

4. Move protocol negotiation out of websocket layer.

  **Pros**: We can get consistent behavior across browsers.
 
 **Cons**: Re-inventing the wheel; solution is not resistant against future protocol changes; binds ActionCable to a content encoding.

Personally I think 1 or 3 are the best.

@jeremy Any opinions on this?

jeremy replied:
#1 is a cool idea, but it also relies on the *server* not returning a `Sec-Websocket-Protocol` choice when `Sec-Websocket-Accept` is omitted.

#2 is the odd one of the bunch.

#3 is nice because it's taking a small, simple step and the cost of fruitless retries is fairly low.

Re. #4… wonder if the client requested
```
Sec-WebSocket-Protocol: actioncable-v1-json, *actioncable-unsupported*
```
Then we can guarantee we'll always have a handshake. From there, the `actioncable-unsupported` protocol's only job is to reject the connection as unsupported.

In any case, seems #3 is best for this PR! Plenty of room to improve in subsequent work 😁

danielrhodes replied:
For 1. This is part of the WebSocket spec. If the server doesn't support any protocols, it needs to not return that header or return the header with a null value. In our case, there is no visibility into this process [as the WebSocket driver does this for us](https://github.com/faye/websocket-driver-ruby/blob/master/lib/websocket/driver/hybi.rb#L76).

Re 4: This is a fantastic idea! I'm going to give this a try. I think this would do the trick.

javan commented on actioncable/app/assets/javascripts/action_cable/events.coffee:
```diff
@@ -0,0 +1,6 @@
+class ActionCable.Events.Disconnected extends ActionCable.Events
+  constructor: (@code, @reason, @isRecoverable) ->
+
+  @ProtocolNotSupported = new Disconnected(4001, "The client and server could not agree on a subprotocol to use.", false)
```
I'm -1 on using a custom Event object to communicate the disconnect info and would prefer a plain object instead.

danielrhodes commented:
Ok I removed the reconnection and replaced it with `actioncable-unsupported`, which works across all browsers and gives us the visibility we need to disconnect if there's a protocol mismatch and tell the app. I'm very happy with that solution. 🎩  tip to @jeremy

maclover7 commented:
This is coming along very nicely -- can you start to squash down some of your commits?

danielrhodes commented:
@maclover7 Yep I plan on doing that before anything is merged!

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,19 +96,28 @@ class ActionCable.Connection
+        @subscriptions.reload()
+      else
+        ActionCable.log("WebSocket got '#{unsupported_protocol}' as the chosen protocol. Stopping monitor and disconnecting.")
+        @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
```
WebSockets are an implementation detail of Action Cable so it smells wrong to dispatch WebSocket-based data here. The only useful data for the client is wether or not AC is going to reconnect automatically or stop.

danielrhodes replied:
Ok yeah I see what you mean: if ActionCable were to support say SSEs or HTTP/2 in the future, reducing the surface area of WebSocket dependencies would be a good thing. :-)

In that case, we should just return an object with a `willAttemptReconnect` property set to `true/false`.

javan replied:
:+1:

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -89,19 +96,28 @@ class ActionCable.Connection
+  disconnect: (event) ->
     return if @disconnected
+    if event and not event.isRecoverable
+      @monitor.stop()
```
Would be nice to add a `Connection#stop` method that closes any existing connection and stops the monitor. It could be called here instead, and would provide an API for clients use (a feature that's been requested). Currently, there isn't an easy way to close the connection because the monitor always reconnects it.

danielrhodes replied:
Check out what I just added. I put the `start` and `stop` methods on the consumer.

kaspth commented on actioncable/app/assets/javascripts/action_cable/consumer.coffee:
```diff
@@ -14,6 +14,17 @@
+#  App.cable.start()
+#
+#  Any channel subscriptions which existed prior to disconnecting will
+#  automatically rescubscribe.
```
Non-standard indentation going on in the added docs. It goes, one for text, three for code.

kaspth replied:
`resubscribe`.

kaspth commented on actioncable/app/assets/javascripts/action_cable/consumer.coffee:
```diff
@@ -22,6 +33,12 @@ class ActionCable.Consumer
+    @connection.open()
+
+  stop: ->
+    @connection.close(true)
```
Passing `true` reads really weird. I think adding a new API could capture this better, perhaps `close(withoutReconnect: true)` could do (if I'm understanding the purpose of the argument right).

javan replied:
Agree re: `true` as an arg, but I'd prefer a `Connection#stop` method instead.

kaspth replied:
Another thing to note is, we're adding extra terminology here. On top of `close` and `open` and the `connected` and `disconnected` callbacks, we now have `start` and `stop`. Since (dis)connect is generally what end-users will see that might be better than start & stop. E.g. `App.cable.connect()` over `App.cable.start()`

danielrhodes replied:
Agreed on the consistency. Changed it.

kaspth commented on actioncable/lib/action_cable.rb:
```diff
@@ -35,7 +37,15 @@ module ActionCable
+    # It is important that the unsupported protocol be last in the list,
+    # as the Websocket Driver always chooses the first protocol that matches,
+    # with the intended result that if the client sends the unsupported
+    # protcol as an option, it will be the last chosen.
```
Could manifest the order requirement by using `protocols.last` in code instead of an extra key. Other parts of the code might even want to stay oblivious to the fact that last protocol means "unsupported", and write it as "whelp, we're all out of protocols, no reason to attempt reconnect."

danielrhodes replied:
Yeah agreed it feels a bit fragile, which is why I was so explicit. I also just added a test to make sure that the protocol is last.

However, instead I could just add in the unsupported protocol right before we instantiate the driver (in `ActionCable::Connection::WebSocket`), which I think would accomplish what you were after when you say
> Other parts of the code might even want to stay oblivious to the fact that last protocol means "unsupported"

This would keep the placement details safe from the rest of the code.

danielrhodes replied:
Ok I did that. Take a look. I think that looks much cleaner and makes it so nobody really has to think about the ordering.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -23,14 +27,16 @@ class ActionCable.Connection
-  close: ->
+  close: ({withoutReconnect} = {}) ->
+    if withoutReconnect is true
+      @monitor.stop()
```
How about:
```coffee
close: ({allowReconnect} = allowReconnect: true) ->
  @monitor.stop() unless allowReconnect
  @webSocket?.close()
```

danielrhodes commented:
@javan I just rolled that entire `disconnect` method into `close` (it's much cleaner this way).

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -77,7 +95,7 @@ class ActionCable.Connection
       switch type
         when message_types.welcome
-          @monitor.recordConnect()
+          break
```
Why was this moved? It introduces a subtle behavior / timing change that doesn't seem relevant to your PR.

danielrhodes replied:
A couple reasons:
1. That `@monitor.recordConnect()` is in a place which makes it protocol dependent. As I recall you wanted to get rid of the welcome message anyways.
2. It was creating a race condition where it was called after the decision to disconnect had been made when there was a protocol mismatch (because the welcome message was sent right after the server opens the connection), but before there was an actual disconnect. Maybe this doesn't matter because the monitor is being shut off anyways, but it seemed like an undesirable behavior.

javan replied:
You sold me on the welcome message in https://github.com/rails/rails/pull/23976#issuecomment-190655306 :grin: 
> I also like the idea that, especially when there is potentially authentication/the possibility that the server will reject the connection, the server sends a definitive message to the client saying you are connected: hence welcome.

danielrhodes replied:
Yes this is true as well, and I just realized there's a weakness I need to fix.

danielrhodes replied:
Your comment got me thinking about terminology, and I guess the question is: what determines a successful connection? And I guess the whole point of this PR is that we regard a successful connection as one where we have a socket connection _and_ we have a successful protocol negotiation. So I put the `recordConnect` back where it was, and also moved the subscription `reload` there as well, because that was a race condition/bound to break.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -23,15 +27,20 @@ class ActionCable.Connection
+    @disconnected = true
+    @monitor.recordDisconnect()
+    @monitor.stop() unless allowReconnect
+    @subscriptions.notifyAll("disconnected", {willAttemptReconnect: allowReconnect})
```
Moving `@monitor.recordDisconnect()` and `@subscriptions.notifyAll()` from the `onclose` handler to the `close` method change their timing such that they don't accurately reflect the connection's state. I also find it confusing that `onclose` calls `close`.

danielrhodes commented:
Ok @javan, changed the disconnection stuff so it's not using the `close` method as much. I think it looks better now.  At least from a timing perspective, everything falls into the right order in the logs.

Either way, it is an improvement from what was there before. There were several methods which "closed" the connection and they were called in different orders at different times. This follows much more logically.

kaspth commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -46,12 +51,21 @@ class ActionCable.Connection
+    @webSocket?.protocol
+
+  isProtocolSupported: ->
+    @getProtocol() in protocols and @getProtocol() isnt unsupported_protocol
```
Still don't understand why we need to pass `unsupported_protocol` explicitly around in all these places. If `protocols` is a by-priority list of protocols for the client to respond to. Then wouldn't on the last protocol really indicate nothing more to negotiate; close the connection altogether. To me that makes more sense than jumping through hoops with this variant of a protocol.

To top it off, I'd like to add the unsupported protocol back to `ActionCable::INTERNAL[:protocols]`, so we can think of it as a manifest of the negotiation order.

danielrhodes replied:
The client isn't choosing the protocol, the server is. The client presents the server with a list of protocols it supports, the server chooses one which matches its own list of supported protocols, and sends that one (or none) back. If the browser client sees that none are sent back, it disconnects. That is all outside of our control.

danielrhodes replied:
> To top it off, I'd like to add the unsupported protocol back to `ActionCable::INTERNAL[:protocols]`, so we can think of it as a manifest of the negotiation order.

Sure I can do this. But on the server side, it does not matter which order they are given to the driver. It only matters that the client give the server the protocols with the unsupported protocol last. So I figured I would keep that ordering logic on the client side.

kaspth replied:
My problem is just that I don't see why we need to do insert the extra protocol at runtime. My understanding is that we can do all that upfront. Maybe that's an insufficient understanding of the domain.

I'll try rereading your comments tomorrow, see if my brain is more receptive then :smile:

Thanks for all your effort on this! This is going to be so cool to have in :heart:

danielrhodes replied:
We don't. I just thought it would be clearer to any future maintainers that this was a special case, but at the same time not get accidentally removed.

danielrhodes commented:
What's the status of this? Are we good to merge? I think after the reviews, the footprint of the change in terms of behavior/code is now quite low.

maclover7 commented:
@danielrhodes I think this is coming along nicely -- can you try to squash down some of your commits?

danielrhodes commented:
Yes will do that now. Github actually just added a feature to do this automatically on merge -- something I'm sure you will appreciate!

javan commented on actioncable/lib/action_cable.rb:
```diff
@@ -35,7 +35,8 @@ module ActionCable
     },
-    default_mount_path: '/cable'.freeze
+    default_mount_path: '/cable'.freeze,
+    protocols: ["actioncable-v1-json".freeze, "actioncable-unsupported".freeze].freeze
```
Apologies for asking again since I think you explained this already, but it's not clear to me what the purpose of actioncable-unsupported protocol is. Perhaps it could be defined as a separate constant that gets included in this array so it can be documented separately.

Additionally, now that we have a v1 protocol in place, what kinds of changes warrant bumping the version? (I know the answer, but think it's worth documenting).

maclover7 replied:
Sorry if I've already asked this -- is there any downside to having the protocol version match the current Action Cable version? Any breaking changes with Action Cable's websockets communication infrastructure should theoretically also be a regular-Action Cable breaking change. Having two version numbers seems a bit much...

danielrhodes replied:
@javan Yeah the reason for this is that the browser side WebSocket client sticks with the bizarre recommended implementation where the client does not tell us why it disconnected, when it could either be a bad disconnect or a protocol mismatch. So the `actioncable-unspported` protocol is the way to get around this: it will always be the last to be chosen, and if it is, we know for sure there is a mismatch. The alternatives to this are really hacky.

@maclover7 I initially asked myself the same thing. However, the protocol version is distinct from the Rails version, and ideally it would be changed quite a bit less often than the Rails version. Also since these protocol changes are almost always breaking changes, semantic versioning is not quite appropriate here.

javan replied:
> **protocols (Optional)**
> Either a single protocol string or an array of protocol strings. These strings are used to indicate sub-protocols, so that a single server can implement multiple WebSocket sub-protocols (for example, you might want one server to be able to handle different types of interactions depending on the specified protocol). If you don't specify a protocol string, an empty string is assumed.
-- https://developer.mozilla.org/en-US/docs/Web/API/WebSocket

Guess I'm still not clear why the client needs know about `actioncable-unsupported`. What would happen if we only pass the supported protocol and let the server disconnect if that protocol isn't specified?

javan replied:
Sorry, that's not right, I know we're not disconnecting server side on protocol mismatch. But can the server set the protocol to `actioncable-unsupported` if the client doesn't open the WebSocket with a supported protocol?

For example, client connects with:
```coffee
@webSocket = new WebSocket(@consumer.url, "actioncable-v1-json")
```
And if the server only knows about `actioncable-v2-json` then it responds by setting the protocol to `actioncable-unsupported`.

danielrhodes replied:
The short answer is no. The server must choose one or none of the protocols the client presents to it. We provide a list of protocols to the Faye-Websocket driver, and we likewise present a list of supported protocols to the browser websocket client. If the server doesn't support any of the protocols, the "correct" way to handle this is to return nothing -- except we get penalized for doing this on the client side because of the implementation recommendations.

The longer answer is that yes we could fork the Faye-Websocket driver and do this... but my experience with the browser websocket client has been that it is extremely stubborn, so it would probably not like this and throw an error if it received a protocol it did not send.

javan replied:
Gotcha. Thanks for explaining!

javan commented on actioncable/app/assets/javascripts/action_cable/subscription.coffee:
```diff
@@ -8,6 +8,12 @@
 #     connected: ->
 #       # Called once the subscription has been successfully completed
 #
+#     disconnected: (event) ->
```
Move away from dispatching an event here. It's a plain object now.

danielrhodes replied:
Are you saying name it something different? Any recommendations?

javan replied:
Yeah, maybe document it as `disconnected: ({ willAttemptReconnect: boolean }) ->`?

javan replied:
I just don't want to suggest that it's an Event instance.

danielrhodes replied:
Yeah makes sense. I just changed it to that.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -23,15 +24,16 @@ class ActionCable.Connection
-    @webSocket?.close()
+  close: ({allowReconnect} = {allowReconnect: true}) ->
+    @monitor.stop() unless allowReconnect
+    @webSocket?.close() unless @isClosed()
```
Could do `@webSocket?.close() if @isActive()` and drop the `isClosed` method entirely. Side note: we intentionally named `isActive` to not match a connection state and `isClosed` doesn't follow that pattern.

danielrhodes replied:
Yep don't see why not. Changed.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -90,18 +103,18 @@ class ActionCable.Connection
       ActionCable.log("WebSocket onclose event")
-      @disconnect()
+      return if @disconnected
+      @disconnected = true
```
Tracking `@disconnected` was originally in place because to prevent calling the `disconnect` method twice since both the `error` and `close` handlers called it. Now that `disconnect` has been removed, I think we can rely on the connection state alone and remove `@disconnected`.

danielrhodes replied:
I think that should stay for now. It still seems to prevent some conditions where `onClose` is called by the websocket even though the websocket is already disconnected, which I think would be confusing to the developer.

javan replied:
:+1: if that's the case. Strange that `close` would be dispatched twice though. Can you reproduce?

danielrhodes replied:
Oh yeah for sure. On every unsuccessful connection this is how the browser says a connection did not happen. I guess it would have made more sense for them to add an event to onError, but what can you do?

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -2,7 +2,7 @@
 # Encapsulate the cable connection held by the consumer. This is an internal class not intended for direct user manipulation.
 
-{message_types} = ActionCable.INTERNAL
+{message_types, protocols} = ActionCable.INTERNAL
```
How about splitting the protocols here into usefully named vars:
```coffee
[supportedProtocols..., unsupportedProtocol] = protocols
````

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -53,6 +58,9 @@ class ActionCable.Connection
   # Private
+  isProtocolSupported: ->
+    [..., unsupported_protocol] = protocols
+    @getProtocol() in protocols and @getProtocol() isnt unsupported_protocol
```
Using the suggested changes above, this would change to:
```coffee
@getProtocol() in supportedProtocols
```

danielrhodes replied:
I had a back and forth with @kaspth about this... he wanted to minimize the heavy separation I made in the code between the supported and unsupported protocols. However, I don't think your suggestion conflicts with what he was looking for, so I made the change.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -74,10 +82,12 @@ class ActionCable.Connection
 
   events:
     message: (event) ->
+      return if @disconnected
```
Is there a scenario where messages are received after calling `webSocket.close()`?

danielrhodes replied:
Yes the server fires out the welcome message pretty fast after connecting. The server doesn't decide to disconnect on a protocol mismatch, the client does.

javan replied:
Can we prevent the server from sending messages on protocol mismatch?

danielrhodes replied:
Yes we could. I ran through some scenarios where we could just tell the server to stop communicating in these cases, but it gets kind of murky and the websocket protocol seems to indicate the onus is on the client to disconnect when it doesn't like the protocol not the server. Also if somebody writes a client in something other than in the browser (that would be me :-P), they might not want that behavior.

javan replied:
Cool. Mostly asking out of curiosity. Adding this guard makes sense.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -53,6 +59,8 @@ class ActionCable.Connection
 
   # Private
+  isProtocolSupported: ->
+    @getProtocol() isnt unsupportedProtocol and @getProtocol() in supportedProtocols
```
Isn't checking `@getProtocol() in supportedProtocols` enough? If the protocol is in that array then it's not `unsupportedProtocol`.

danielrhodes replied:
Sure. It's just extra safety I guess.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -90,18 +100,18 @@ class ActionCable.Connection
     open: ->
       ActionCable.log("WebSocket onopen event")
       @disconnected = false
```
Should we set `disconnected = false` only if the protocol is supported and remove the `disconnected` check in the `onMessage` handler?

danielrhodes replied:
I tried this before, but that welcome message was coming in and making the connection monitor think it was still online. It doesn't happen when code is as it is.

javan replied:
```coffee
open: ->
  ActionCable.log("WebSocket onopen event, protocol is '#{@getProtocol()}'")
  if @isProtocolSupported()
    @disconnected = false
  else
    ActionCable.log("WebSocket opened with an unsupported protocol. Stopping monitor and disconnecting.")
    @close(allowReconnect: false)
```

javan replied:
Guess that ^ doesn't remove the need to check `disconnected` in the message handler, but I find it clearer since it consolidates the logging and removes the early `return`.

danielrhodes replied:
Yeah I was fooling around with this and was going back and forth, but then I settled on a successful connection means actual server connection _and_ a protocol is chosen. That work we did with the welcome message was actually quite useful in telling the client something was wrong during authentication failure, and we would lose this in the case where a successful connection means just connecting to the server.

danielrhodes replied:
Also I found a bug that hadn't been caught before which was that subscription requests were fired off immediately on connecting, even if the server had not sent welcome yet. In some scenarios that would produce odd behavior I imagine.

javan replied:
We've hit sporadic missing subscriptions in Basecamp. Hoping this is the cause! :heart:

javan replied:
>  I settled on a successful connection means actual server connection and a protocol is chosen.

:+1: to that, but you're setting `disconnected = false` before checking the protocol.

javan replied:
Reviewing the `open` handler and where it sets `@disconnected` is the only thing left to address before this can be merged, IMO.

danielrhodes replied:
Alright I changed the message handler to check `@isActive` instead. It works the same. I don't quite understand why it worked with `@disconnected` even though as you pointed out there could be a race condition -- whatever underlying threading is going on obviously prioritized the `close` before the `message` is all I can surmise.

jeremy replied:
Re. subscription requests before welcome: we [buffer incoming messages and process the buffer after welcome](https://github.com/rails/rails/blob/master/actioncable/lib/action_cable/connection/base.rb#L168-L170), so sending early subscription requests should work as expected.

javan commented:
Looking great! Any issues deploying the server-side changes while clients still have the previous .js assets running?

danielrhodes commented:
@javan There shouldn't be. If the client doesn't send any protocols it wants to use, the server won't either.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -53,6 +59,8 @@ class ActionCable.Connection
     @isState("open", "connecting")
 
   # Private
```
Sorry to nitpick, but add a newline after this comment please. Indicates that all methods below are "private", not just the `isPorotocolSupported` method.

danielrhodes replied:
:+1:

danielrhodes commented:
Squashed the commits for squashlover7, err I mean @maclover7 :-P

P.S. I've actually started to like squashing now too, even if it's a conspiracy to keep everybody's commit numbers low to suppress rankings.

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -74,10 +83,12 @@ class ActionCable.Connection
 
   events:
     message: (event) ->
+      return if not @isActive
```
This isn't right. `@isActive` is a function, not a property.

I *think* we want to check that the WebSocket is using a supported protocol here so messages aren't received during the brief window after successfully connecting using an unsupported protocol.
```coffee
message: (event) ->
  return unless @isSupportedProtocol()
  ...
```

danielrhodes replied:
:+1:

javan commented on actioncable/app/assets/javascripts/action_cable/connection.coffee:
```diff
@@ -90,18 +101,18 @@ class ActionCable.Connection
+        ActionCable.log("WebSocket got '#{@getProtocol()}' as the chosen protocol. Stopping monitor and disconnecting.")
+        @close(allowReconnect: false)
+        return
+      ActionCable.log("WebSocket using '#{@getProtocol()}' subprotocol.")
```
Not loving how this method reads now. The early `return` only guards against logging. I suggest:
```coffee
open: ->
  ActionCable.log("WebSocket onopen event, using '#{@getProtocol()}' subprotocol")
  @disconnected = false
  unless @isProtocolSupported()
    ActionCable.log("Protocol is unsupported. Stopping monitor and disconnecting.")
    @close(allowReconnect: false)
```

danielrhodes replied:
:+1:

javan commented:
OK to merge as far as I'm concerned. Thanks for sticking with this one @danielrhodes!

jeremy commented:
Fantastic work @danielrhodes!

kaspth commented:
Well done! ❤️

cdurante commented:
Just re-bundled with this PR merged, previously working AC web socket connections now throw browser js error on  Safari Version 9.1 (11601.5.17.1)?


``
[Error] TypeError: this.isSupportedProtocol is not a function. (In 'this.isSupportedProtocol()', 'this.isSupportedProtocol' is undefined)
	message (connection.self-7605b1bcff260663a68e0847ec3cfece34a28e2cb3c61fb59bcac2dfab6b6ad7.js:131)
	(anonymous function)
``

jeremy commented:
[Sure is](https://github.com/rails/rails/commit/9c2945f4cc1017fc9424f2e425e4820da1ef33b0). Thanks @cdurante!
